### PR TITLE
adds available_device to RocCurve #3335

### PR DESCRIPTION
**Pull request by BanzaiTokyo**

### CODE REVIEW

File: tests/ignite/metrics/test_roc_curve.py [hunk 1]
```diff
@@ -49,9 +49,11 @@ def test_roc_curve(available_device):
     sk_fpr, sk_tpr, sk_thresholds = roc_curve(
         to_float32_if_mps(np_y, available_device), to_float32_if_mps(np_y_pred, available_device)
     )
-    sk_fpr = to_float32_if_mps(sk_fpr, available_device)
-    sk_tpr = to_float32_if_mps(sk_tpr, available_device)
-    sk_thresholds = to_float32_if_mps(sk_thresholds, available_device)
+
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:52]: Let's create np_y_pred and np_y with torch.rand and torch.zeros in this test as well

Comment (vfdev-5) [reply tests/ignite/metrics/test_roc_curve.py:57]: @BanzaiTokyo check this comment ^^^

File: tests/ignite/metrics/test_roc_curve.py [hunk 2]
```diff
@@ -48,8 +48,8 @@ def test_no_sklearn(mock_no_sklearn):
 
 def test_roc_curve(available_device):
     size = 100
-    np_y_pred = np.random.rand(size, 1)
-    np_y = np.zeros((size,))
+    np_y_pred = torch.rand(size, 1).numpy()
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:51]: Use torch tensors instead of numpy arrays

File: tests/ignite/metrics/test_roc_curve.py [hunk 3]
```diff
@@ -47,14 +34,15 @@ def test_no_sklearn(mock_no_sklearn):
 
 
 def test_roc_curve(available_device):
+
+    dtype = torch.float32 if available_device == "mps" else torch.get_default_dtype()
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:38]: We do not need this

File: docs/source/metrics.rst [hunk 4]
```diff
@@ -283,7 +283,7 @@ By default, `Metrics` are epoch-wise, it means
 - :meth:`~ignite.metrics.metric.Metric.update()` is triggered every ``ITERATION_COMPLETED``.
 - :meth:`~ignite.metrics.metric.Metric.compute()` is triggered every ``EPOCH_COMPLETED``.
 
-Usages can be user defined by creating a class inheriting for :class:`~ignite.metrics.metric.MetricUsage`. See the list below of usages.
+Usages can be user defined by creating a class inheriting from :class:`~ignite.metrics.metric.MetricUsage`. See the list below of usages.
```
Comment (vfdev-5) [inline docs/source/metrics.rst:286]: Something wrong with your rebase. This is already on master

File: ignite/metrics/roc_auc.py [hunk 5]
```diff
@@ -198,9 +198,10 @@ def compute(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:  # type: i
         if idist.get_rank() == 0:
             # Run compute_fn on zero rank only
             fpr, tpr, thresholds = cast(Tuple, self.compute_fn(_prediction_tensor, _target_tensor))
-            fpr = torch.tensor(fpr, device=_prediction_tensor.device)
-            tpr = torch.tensor(tpr, device=_prediction_tensor.device)
-            thresholds = torch.tensor(thresholds, device=_prediction_tensor.device)
+            dtype = torch.float32 if _prediction_tensor.device.type == "mps" else torch.get_default_dtype()
```
Comment (vfdev-5) [inline ignite/metrics/roc_auc.py:201]: Use `self._double_dtype` which is defined in Metric instead of this

File: tests/ignite/metrics/test_roc_curve.py [hunk 6]
```diff
@@ -33,104 +33,128 @@ def test_no_sklearn(mock_no_sklearn):
...
-    assert np.array_equal(tpr, sk_tpr)
+    fpr = fpr.cpu().numpy()
+    tpr = tpr.cpu().numpy()
+    thresholds = thresholds.cpu()
+
+    assert np.allclose(fpr, expected_fpr, rtol=1e-6)
+    assert np.allclose(tpr, expected_tpr, rtol=1e-6)
     # assert thresholds almost equal, due to numpy->torch->numpy conversion
-    np.testing.assert_array_almost_equal(thresholds, sk_thresholds)
+    np.testing.assert_array_almost_equal(thresholds, expected_thresholds)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:59]: Let's try to do 
```
assert expected_fpr == pytest.approx(fpr)
...
```

Comment (BanzaiTokyo) [reply tests/ignite/metrics/test_roc_curve.py:56]: are you sure? From what I read:
```

Comparing small floats or scalars | pytest.approx()
-- 
Comparing arrays with fine-tuned precision | np.allclose()
--
Comparing arrays and want good error output | np.testing.assert_array_almost_equal()
```

Comment (vfdev-5) [reply tests/ignite/metrics/test_roc_curve.py:56]:
```
>>> import torch
>>> import numpy as np
>>> import pytest
>>> np.arange(100) == pytest.approx(torch.arange(100))
True
```

File: tests/ignite/metrics/test_roc_curve.py [hunk 7]
```diff
@@ -33,104 +33,115 @@ def test_no_sklearn(mock_no_sklearn):
...
 
-def test_integration_roc_curve_with_output_transform():
+def test_integration_roc_curve_with_output_transform(available_device):
     np.random.seed(1)
     size = 100
-    np_y_pred = np.random.rand(size, 1)
-    np_y = np.zeros((size,))
-    np_y[size // 2 :] = 1
-    np.random.shuffle(np_y)
+    y_pred = torch.rand(size, 1)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:61]: Can you also remove `np.random.seed(1)` everywhere

File: tests/ignite/metrics/test_roc_curve.py [hunk 8]
```diff
@@ -33,104 +33,115 @@ def test_no_sklearn(mock_no_sklearn):
...
+    y = torch.zeros(size)
+    y[size // 2 :] = 1
+    np.random.shuffle(y)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:64]: Let's use `torch.randint(0, 2, size=(size,))`

File: tests/ignite/metrics/test_roc_curve.py [hunk 9]
```diff
@@ -33,104 +32,109 @@ def test_no_sklearn(mock_no_sklearn):
...
-def test_roc_curve():
+def test_roc_curve(available_device):
     size = 100
-    np_y_pred = np.random.rand(size, 1)
-    np_y = np.zeros((size,))
-    np_y[size // 2 :] = 1
-    sk_fpr, sk_tpr, sk_thresholds = roc_curve(np_y, np_y_pred)
+    y_pred = torch.rand(size, 1)
+    y = torch.zeros(size)
+    y[size // 2 :] = 1
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:39]:
```suggestion
    y = torch.randint(0, 2, size=(size,))
```

File: tests/ignite/metrics/test_roc_curve.py [hunk 10]
```diff
@@ -165,8 +168,8 @@ def update(engine, i):
 
     y = idist.all_gather(y)
     y_pred = idist.all_gather(y_pred)
-    sk_fpr, sk_tpr, sk_thresholds = roc_curve(y.cpu().numpy(), y_pred.cpu().numpy())
+    expected_fpr, expected_tpr, expected_thresholds = roc_curve(y.cpu().numpy(), y_pred.cpu().numpy())
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_roc_curve.py:171]:
```suggestion
    expected_fpr, expected_tpr, expected_thresholds = roc_curve(y, y_pred)
```
